### PR TITLE
#24 verify email address

### PR DESCRIPTION
**Pull request by AndresAusecha**

### CODE REVIEW

File: user-profile/src/main/kotlin/systemkern/profile/configuration.kt [hunk 1]
```diff
@@ -34,6 +35,12 @@ internal class CustomWebSecurityConfigurerAdapter(
     val pattern1: String = "/user-profiles/"
     val pattern2: String = "/user-profiles/{\\d+}"
 
+    @Throws(Exception::class)
+    override fun configure(webSecurity: WebSecurity) {
+        webSecurity
+            .ignoring()
+            .antMatchers(HttpMethod.PUT,"$pattern/verify-email")
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/configuration.kt]: please add a static import here

File: user-profile/src/main/kotlin/systemkern/profile/UserProfileRepository.kt [hunk 2]
```diff
@@ -62,6 +101,8 @@ internal class SessionListener : HttpSessionListener {
     }
 
     override fun sessionCreated(event: HttpSessionEvent) {
-        event.session.maxInactiveInterval =  Parameters.sessionTime.toInt() * 60
+        event.session.maxInactiveInterval = Parameters.sessionTime.toInt() * 60
     }
 }
+@ResponseStatus(HttpStatus.NOT_ACCEPTABLE)
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/UserProfileRepository.kt]: please add a static import here as well

File: user-profile/src/main/kotlin/systemkern/profile/UserProfileRepository.kt [hunk 3]
```diff
@@ -7,25 +7,65 @@ import org.springframework.data.repository.CrudRepository
...
+    val email: String,
+    val verified: Boolean
+)
+
 @Component
-internal class UserProfileEntiyListener(
+internal class UserProfileEntityListener(
     internal val passwordEncoder: BCryptPasswordEncoder = BCryptPasswordEncoder()
 ) {
+    val pattern = Pattern.compile("^[_A-Za-z0-9-\\+]+(\\.[_A-Za-z0-9-]+)*" +
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/UserProfileRepository.kt]: please rename to `emailAddressPattern` and add a unit test for the pattern

File: user-profile/src/main/kotlin/systemkern/profile/UserProfileRepository.kt [hunk 4]
```diff
@@ -7,25 +7,65 @@ import org.springframework.data.repository.CrudRepository
...
+import java.time.LocalDateTime
 import java.util.*
 import javax.persistence.*
 import javax.servlet.http.HttpSessionEvent
 import javax.servlet.http.HttpSessionListener
+import java.util.regex.Pattern
+import java.util.zip.DataFormatException
+
+@RestController
+internal class UserProfileController(val repo1: UserProfileRepository,
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/UserProfileRepository.kt]: please use usable property names

for example
`repo` and `emailRepo`

File: user-profile/src/main/kotlin/systemkern/profile/EmailVerificationRepository.kt [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import java.time.LocalDateTime
+import java.util.*
+import javax.persistence.Entity
+import javax.persistence.Id
+
+@RepositoryRestResource
+@Component
+internal interface EmailVerificationRepository : CrudRepository<EmailVerification, UUID> {
+    @PostMapping("verify-email/{tokenId}")
+    @Query("SELECT * FROM public.user_profile u WHERE u.id = :tokenId", nativeQuery = true)
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/EmailVerificationRepository.kt]: PLease add a JPA test for this query

File: integration-tests/src/test/kotlin/systemkern/profile/UserControllerIT.kt [hunk 6]
```diff
@@ -81,16 +82,28 @@ internal class UserControllerIT : IntegrationTest() {
...
-            ))
+                responseFields(listOf(
+                    fieldWithPath("url").description("Url to verify user email").type(STRING))
+                )))
+            .andReturn().response.contentAsString.let { urlToVerifyUserProfile = "url" +
+                    JSONObject(it).get("url").toString() }
+        `verify email`(user.username,user.password)
+    }
+
+    private fun `verify email`(username: String, password: String){
```
Comment (rainerkern) [inline integration-tests/src/test/kotlin/systemkern/profile/UserControllerIT.kt]: Hi Andres,
please be careful when using these kind of function names with spaces and `'`, It makes the code harder to read.

My thinking is to use it only for test functions. There it makes sense because the names are printed into the console and the IntelliJ UI ... this makes reading them actually easier. But like this in the code I think it is harder.

File: runtime-cli/pom.xml [hunk 7]
```diff
@@ -53,6 +53,11 @@
             <groupId>org.postgresql</groupId>
             <artifactId>postgresql</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-mail</artifactId>
+            <version>1.4.3.RELEASE</version>
```
Comment (rainerkern) [inline runtime-cli/pom.xml]: please ! ... no version numbers in the module poms ... especially not ones that are managed by the Spring Boot POM

File: user-profile/pom.xml [hunk 8]
```diff
@@ -53,6 +53,17 @@
             <artifactId>spring-integration-core</artifactId>
             <version>${spring-integration-core}</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-context-support</artifactId>
+            <version>5.1.0.BUILD-SNAPSHOT</version>
```
Comment (rainerkern) [inline user-profile/pom.xml]: please ! ... no version numbers in the module poms ... especially not ones that are managed by the Spring Boot POM

File: user-profile/pom.xml [hunk 9]
```diff
@@ -53,6 +53,17 @@
...
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-beans</artifactId>
+            <version>5.1.0.BUILD-SNAPSHOT</version>
```
Comment (rainerkern) [inline user-profile/pom.xml]: please ! ... no version numbers in the module poms ... especially not ones that are managed by the Spring Boot POM

File: user-profile/src/main/kotlin/systemkern/profile/AuthenticationController.kt [hunk 10]
```diff
@@ -17,24 +15,7 @@ internal class AuthenticationController(
     internal fun login(auth: Authentication,
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/AuthenticationController.kt]: formatting should be either in one line
```
internal fun login(auth: Authentication, @RequestHeader password: String): AuthenticationResponse {
```

or like this
```
internal fun login(
    auth: Authentication,
    @RequestHeader password: String
): AuthenticationResponse {
```

File: user-profile/src/main/kotlin/systemkern/profile/AuthenticationService.kt [hunk 11]
```diff
@@ -1,12 +1,16 @@
...
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
 import org.springframework.stereotype.Service
+import java.time.LocalDateTime
 import java.util.*
 import javax.servlet.http.HttpServletRequest
 import kotlin.collections.HashMap
 
 @Service
-internal class AuthenticationService(val repo: UserProfileRepository) {
+internal class AuthenticationService(val repo: UserProfileRepository,
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/AuthenticationService.kt]: formatting

File: user-profile/src/main/kotlin/systemkern/profile/AuthenticationService.kt [hunk 12]
```diff
@@ -28,4 +32,25 @@ internal class AuthenticationService(val repo: UserProfileRepository) {
     internal fun deleteToken(token: UUID) {
         tokens.remove(token)
     }
+
+    @Throws(UserNotFoundException::class)
+    internal fun authenticationProcess(auth: Authentication,
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/AuthenticationService.kt]: formatting

File: user-profile/src/main/kotlin/systemkern/profile/AuthenticationService.kt [hunk 13]
```diff
@@ -28,4 +32,25 @@ internal class AuthenticationService(val repo: UserProfileRepository) {
...
+                                       password: String
+    ): AuthenticationResponse {
+
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/AuthenticationService.kt]: no empty line here

File: user-profile/src/main/kotlin/systemkern/profile/UserProfileRepository.kt [hunk 14]
```diff
@@ -104,5 +132,6 @@ internal class SessionListener : HttpSessionListener {
         event.session.maxInactiveInterval = Parameters.sessionTime.toInt() * 60
     }
 }
+
 @ResponseStatus(HttpStatus.NOT_ACCEPTABLE)
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/UserProfileRepository.kt]: please add a static import for `HttpStatus.NOT_ACCEPTABLE`

File: user-profile/src/main/kotlin/systemkern/profile/configuration.kt [hunk 15]
```diff
@@ -59,6 +58,9 @@ internal class CustomWebSecurityConfigurerAdapter(
             .antMatchers(HttpMethod.GET, pattern, pattern1)
             .denyAll()
 
+            .antMatchers(HttpMethod.POST,"/verify-email/{\\d+}")
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/configuration.kt]: please add an import for `HttpMethod.POST`

File: user-profile/src/main/kotlin/systemkern/profile/configuration.kt [hunk 16]
```diff
@@ -59,6 +58,9 @@ internal class CustomWebSecurityConfigurerAdapter(
             .antMatchers(HttpMethod.GET, pattern, pattern1)
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/configuration.kt]: please add an import for `HttpMethod.POST`

File: runtime-cli/pom.xml (hunk 7 above)
Comment (AndresAusecha) [reply runtime-cli/pom.xml]: Yeap, dont worry about the POM's. I will review them in the last momento before to merge.

**Comment by rainerkern** [comment]
Hi Andres, good morning.

Can you please first fix the merge conflicts in this branch before I review it.
This is usually done by merging from master -> this branch

cheers
Rainer

**Comment by AndresAusecha** [comment]
OK i will try

File: .travis.yml [hunk 17]
```diff
@@ -7,4 +7,27 @@ services:
   - docker
 
 # -B batch mode makes Maven less verbose
-script: mvn -B clean verify
+# Useles change
+<<<<<<< HEAD
```
Comment (rainerkern) [inline .travis.yml]: Andres ... this is not how to resolve merge conflicts :(

File: CONTRIBUTE.md [hunk 18]
```diff
@@ -13,14 +13,31 @@ code-style.kotlin.tabs-and-indents.continuation-indent = 4
 
 ## How to submit the Code - Workflow
 1.  Create a personal fork of the project on Github and clone it locally.
+<<<<<<< HEAD
```
Comment (rainerkern) [inline CONTRIBUTE.md]: Please revert these changes

File: integration-tests/pom.xml [hunk 19]
```diff
@@ -13,7 +13,7 @@
     <artifactId>integration-tests</artifactId>
 
     <properties>
-        <org.springframework.restdocs.spring-restdocs-asciidoctor.version>2.0.1.RELEASE</org.springframework.restdocs.spring-restdocs-asciidoctor.version>
+        <org.springframework.restdocs.spring-restdocs-asciidoctor.version>1.2.2.RELEASE</org.springframework.restdocs.spring-restdocs-asciidoctor.version>
```
Comment (rainerkern) [inline integration-tests/pom.xml]: why the change to the much older version?

File: parent/pom.xml [hunk 20]
```diff
@@ -20,11 +20,19 @@
         <java.version>1.8</java.version>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-        <org.apache.maven.wagon.wagon-ftp.version>3.1.0</org.apache.maven.wagon.wagon-ftp.version>
+        <org.apache.maven.wagon.wagon-ftp.version>1.0-beta-6</org.apache.maven.wagon.wagon-ftp.version>
 
         <!-- dependency versions-->
         <com.nhaarman.mockito-kotlin.version>1.5.0</com.nhaarman.mockito-kotlin.version>
         <org.javamoney.moneta.version>1.1</org.javamoney.moneta.version>
+        <org.zalando.jackson-datatype-money.version>1.0.0-RC2</org.zalando.jackson-datatype-money.version>
```
Comment (rainerkern) [inline parent/pom.xml]: please check which of those dependencies are managed by spring boot
I am quite shure that at least the following are managed by spring boot
* spring-integration
* spring-context
* spring-beans
* spring-start-mail
* hivernate-core

File: user-profile/pom.xml [hunk 21]
```diff
@@ -50,7 +50,37 @@
         <dependency>
             <groupId>org.springframework.security</groupId>
             <artifactId>spring-security-web</artifactId>
+<<<<<<< HEAD
```
Comment (rainerkern) [inline user-profile/pom.xml]: there is a merge error here

File: user-profile/src/main/kotlin/systemkern/profile/MailUtility.kt [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.context.annotation.Configuration
+import org.springframework.mail.javamail.JavaMailSenderImpl
+import org.springframework.web.servlet.config.annotation.EnableWebMvc
+
+
+@Configuration
+@EnableWebMvc
+@ComponentScan(basePackages = ["systemkern.profile"])
+class WebConfig {
+    //CODE CODE CODE
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/MailUtility.kt]: ?

File: user-profile/src/main/kotlin/systemkern/profile/MailUtility.kt [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+        javaMailSender.host = "localhost"
+        javaMailSender.port = 1025
+
+        return javaMailSender
+    }
+
+}
+
+@Component
+internal class MailUtility(@Autowired
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/MailUtility.kt]: @Autowired is not necessary in Kotlin

File: user-profile/src/main/kotlin/systemkern/profile/MailUtility.kt [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+                           private val emailSender: JavaMailSender
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/MailUtility.kt]: indentation

File: user-profile/src/main/kotlin/systemkern/profile/MailUtility.kt [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+) {
+    internal var urlToVerify: String = ""
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/MailUtility.kt]: why is `urlToVerify` a var
and why is it not in the constructor

File: user-profile/src/main/kotlin/systemkern/profile/UserProfile.kt [hunk 26]
```diff
@@ -5,17 +5,20 @@ import com.fasterxml.jackson.annotation.JsonProperty.Access.WRITE_ONLY
...
+    @Column(name = "id_user_profile")
+    val id_userProfile: UUID = randomUUID(),
     val name: String,
 
     @JsonProperty(access = WRITE_ONLY)
     var password: String, /*This attribute is var because of how repository event handler works*/
-    val username: String
+    val username: String,
+    val email: String,
+    @OneToMany(fetch = FetchType.LAZY)
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/UserProfile.kt]: please make a static import for the FetchType.LAZY

File: user-profile/src/main/kotlin/systemkern/profile/configuration.kt [hunk 27]
```diff
@@ -34,24 +38,33 @@ internal class CustomWebSecurityConfigurerAdapter(
...
+        webSecurity
+            .ignoring()
+    }
+
     @Throws(Exception::class)
     override fun configure(http: HttpSecurity) {
         http.csrf()
             .disable()
             .authorizeRequests()
-            .antMatchers(HttpMethod.DELETE, pattern, pattern1, pattern2)
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/configuration.kt]: 👍

File: user-profile/src/main/kotlin/systemkern/profile/UserProfile.kt [hunk 28]
```diff
@@ -5,17 +5,20 @@ import com.fasterxml.jackson.annotation.JsonProperty.Access.WRITE_ONLY
...
 
 @Entity
-@EntityListeners(UserProfileEntiyListener::class)
+@EntityListeners(UserProfileEntityListener::class)
 internal data class UserProfile(
     @Id
-    @GeneratedValue(strategy = AUTO)
-    val id: UUID = randomUUID(),
+    @Column(name = "id_user_profile")
+    val id_userProfile: UUID = randomUUID(),
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/UserProfile.kt]: multiple issues here
1. please leave the name for id as 'id'
2. `id_userProfile` is not a Java/Kotlin variable name
3. the @Column annotation is redundant

File: integration-tests/pom.xml (hunk 19 above)
Comment (AndresAusecha) [reply integration-tests/pom.xml]: I was trying every-thing to fix the errors

File: runtime-cli/src/main/resources/application.properties [hunk 29]
```diff
@@ -18,7 +18,7 @@ server.port=8080
 spring.datasource.url=jdbc:postgresql://localhost:5432/postgres
 spring.datasource.username=postgres
 spring.datasource.password=postgres
-spring.jpa.hibernate.ddl-auto=create
```
Comment (rainerkern) [inline runtime-cli/src/main/resources/application.properties]: 👍

File: user-profile/pom.xml [hunk 30]
```diff
@@ -57,36 +57,36 @@
...
             <artifactId>spring-integration-core</artifactId>
-            <version>${spring-integration-core}</version>
-=======
->>>>>>> master
         </dependency>
 
         <dependency>
             <groupId>org.springframework</groupId>
             <artifactId>spring-context-support</artifactId>
-            <version>${spring-context-support-ver}</version>
```
Comment (rainerkern) [inline user-profile/pom.xml]: 👍

File: user-profile/src/main/kotlin/systemkern/profile/EmailVerification.kt [hunk 31]
```diff
@@ -42,15 +45,14 @@ internal interface EmailVerificationRepository : CrudRepository<EmailVerificatio
...
 internal data class EmailVerification(
     @Id
-    @Column(name = "id_email_verification")
     val id_email_verification: UUID,
     val creationDate: LocalDateTime,
     val validUntil: LocalDateTime,
     var completionDate: LocalDateTime,
-    @ManyToOne(fetch = FetchType.LAZY)
+    @ManyToOne(fetch = LAZY)
     @JoinColumn(name = "id_user_profile", nullable = false)
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/EmailVerification.kt]: sure that this is still correct?

File: user-profile/src/main/kotlin/systemkern/profile/MailUtility.kt [hunk 32]
```diff
@@ -1,41 +1,15 @@
...
-
-        return javaMailSender
-    }
-
-}
 
 @Component
-internal class MailUtility(@Autowired
-                           private val emailSender: JavaMailSender
+internal class MailUtility(private val emailSender: JavaMailSender,
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/MailUtility.kt]: almost :-)
please put email sender in a seaparate line

File: user-profile/src/main/kotlin/systemkern/profile/configuration.kt [hunk 33]
```diff
@@ -116,21 +110,21 @@ internal class AuthenticationFilter(
...
-    private fun procUsernamePasswordAuth(
-        request: HttpServletRequest,
-        httpResponse: HttpServletResponse,
-        username: String,
-        password: String
-    ) {
+    private fun procUsernamePasswordAuth(request: HttpServletRequest,
+                                         httpResponse: HttpServletResponse,
+                                         username: String,
+                                         password: String) {
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/configuration.kt]: please put the `) {`on a separate line

File: user-profile/src/main/kotlin/systemkern/profile/configuration.kt [hunk 34]
```diff
@@ -116,21 +110,21 @@ internal class AuthenticationFilter(
...
         filter.doFilter(request, response)
     }
 
-    private fun procUsernamePasswordAuth(
-        request: HttpServletRequest,
-        httpResponse: HttpServletResponse,
-        username: String,
-        password: String
-    ) {
+    private fun procUsernamePasswordAuth(request: HttpServletRequest,
```
Comment (rainerkern) [inline user-profile/src/main/kotlin/systemkern/profile/configuration.kt]: please put the request parameter on a separate line

File: user-profile/src/test/resources/application-integration-test.properties [hunk 35]
```diff
@@ -8,11 +8,8 @@
...
-hibernate.connection.driver_class"=org.postgresql.Driver
+spring.jpa.hibernate.ddl-auto=create
 
 # Email configurations
-spring.mail.host=localhost
-spring.mail.port=1025
-spring.mail.properties.mail.smtp.auth=false
-spring.mail.properties.mail.smtp.starttls.enable=false
\ No newline at end of file
+spring.mail.host=192.168.99.100
```
Comment (rainerkern) [inline user-profile/src/test/resources/application-integration-test.properties]: Is this IP correct ... i think it should be localhost.
will this work on travis?
